### PR TITLE
GODRIVER-1530 Deprecate topology.WithConnString, which duplicates logic

### PR DESCRIPTION
**Pull request by prestonvasquez**
GODRIVER-1530
GODRIVER-1760

There is a near line-for-line duplication of logic between `client.configure` and `topology.WithConnString`.

The goal of this PR is to deprecate the functional options pattern (including `topology.WithConnString`) for constructing a topology in favor of passing a user-built `topology.Config` struct into the topology constructor, like so:

```go
topo, err := topology.New(&topology.Config{ ... })
```

To make the construction more natural for any potential external use cases, this POC also seeks to construct a `topology.Config` object from an `options.ClientOption`:

```go
cfg, _ := topology.NewConfig(options.Client().ApplyURI(uri))
topo, err := topology.New(cfg)
```

Finally, to provide a methodology of setting and accessing unexported data on the stable client internally, there is also a way to construct a config with a client implementation. 

```go
// topology package
type client interface {
  SetServerAPI(*driver.ServerAPIOptions)
}

func NewConfigWithClient(co options.ClientOptions, client client) {
  // ... some logic 
  client.SetServerAPI(serverAPI)
}

// stable mongo package
type topologyClient struct{ *Client }

func (tc *topologyClient) SetServerAPI(serverAPI *driver.ServerAPIOptions) {
  tc.Client.serverAPI = serverAPI
}

func New(opts options.ClientOptions) *Client {
  // ... some logic
  cfg, _ := topology.NewConfigWithClient(opts, &topologyClient{client})
  // create deployment ...
}
```
This pattern will allow us to set data on the client without exporting information in the stable API. Together, these updates will allow us to move the logic from `client.configure` into the topology package and remove the functional options pattern for constructing a topology config.

### CODE REVIEW

File: x/mongo/driver/topology/topology.go [hunk 1]
```diff
@@ -119,6 +119,32 @@ func newServerSelectionState(selector description.ServerSelector, timeoutChan <-
 	}
 }
 
+func New_(cfg *config) (*Topology, error) {
```
Comment (prestonvasquez) [inline x/mongo/driver/topology/topology.go:122]: This will fully replace `topology.New` once we've determine how to handle some edge case test conversions.

File: mongo/options/clientoptions_test.go [hunk 2]
```diff
@@ -619,7 +619,6 @@ func TestClientOptions(t *testing.T) {
 			{"multiple hosts in options", Client().SetHosts([]string{"foo", "bar"}), internal.ErrLoadBalancedWithMultipleHosts},
 			{"replica set name", Client().SetReplicaSet("foo"), internal.ErrLoadBalancedWithReplicaSet},
 			{"directConnection=true", Client().SetDirect(true), internal.ErrLoadBalancedWithDirectConnection},
-			{"directConnection=false", Client().SetDirect(false), internal.ErrLoadBalancedWithDirectConnection},
```
Comment (prestonvasquez) [inline mongo/options/clientoptions_test.go:622]: I don't think this is needed based on [the spec](https://github.com/mongodb/specifications/blob/8bd2af8caf670289b60df01e892a112509975d52/source/load-balancers/load-balancers.rst#:~:text=The%20connection%20string%20contains%20a%20directConnection%20option%20with%20a%20value%20of%20true.). I can't find anything that specifies that a `false` direct connection can't be load balanced.

File: mongo/options/clientoptions.go [hunk 3]
```diff
@@ -204,7 +204,7 @@ func (c *ClientOptions) validate() error {
 		if c.ReplicaSet != nil {
 			return internal.ErrLoadBalancedWithReplicaSet
 		}
-		if c.Direct != nil {
+		if c.Direct != nil && *c.Direct {
```
Comment (prestonvasquez) [inline mongo/options/clientoptions.go:207]: Based on [this](https://github.com/mongodb/specifications/blob/8bd2af8caf670289b60df01e892a112509975d52/source/load-balancers/load-balancers.rst#:~:text=The%20connection%20string%20contains%20a%20directConnection%20option%20with%20a%20value%20of%20true.) spec , it looks like we only need to error when direct is specifically `true`.

File: x/mongo/driver/topology/topology_options_test.go [hunk 4]
```diff
@@ -9,73 +9,49 @@ package topology
...
 	"fmt"
 	"testing"
-	"time"
 
 	"github.com/stretchr/testify/assert"
-	"go.mongodb.org/mongo-driver/x/mongo/driver/connstring"
+	"go.mongodb.org/mongo-driver/mongo/options"
 )
 
-func TestOptionsSetting(t *testing.T) {
```
Comment (prestonvasquez) [inline x/mongo/driver/topology/topology_options_test.go:18]: I don't think this test is needed anymore since we no longer support functional config options.

File: x/mongo/driver/topology/topology_options_test.go [hunk 5]
```diff
@@ -9,73 +9,49 @@ package topology
...
-	}
-	for _, tc := range testCases {
-		t.Run(tc.name, func(t *testing.T) {
-			topo, err := New(WithConnString(func(connstring.ConnString) connstring.ConnString { return tc.cs }))
-			assert.Nil(t, err, "topology.New error: %v", err)
-			assert.Equal(t, tc.mode, topo.cfg.mode, "expected mode %v, got %v", tc.mode, topo.cfg.mode)
-		})
-	}
-}
+//func TestDirectConnectionFromConnString(t *testing.T) {
```
Comment (prestonvasquez) [inline x/mongo/driver/topology/topology_options_test.go:17]: Is this test still needed? I'm not sure how exactly it would work with the new API.

File: x/mongo/driver/topology/topology.go [hunk 6]
```diff
@@ -142,15 +143,20 @@ func New(opts ...Option) (*Topology, error) {
...
+		t.pollingRequired = strings.HasPrefix(t.cfg.URI, "mongodb+srv://") && !t.cfg.LoadBalanced
 	}
 
 	t.publishTopologyOpeningEvent()
 
 	return t, nil
 }
 
+// NewDefault creates a new unconfigured topolgoy.
+func NewDefault() (*Topology, error) {
```
Comment (prestonvasquez) [inline x/mongo/driver/topology/topology.go:156]: This could be dropped in favor of using `New(nil)` inline, though this does look a little strange.

File: x/mongo/driver/topology/topology_options.go [hunk 7]
```diff
@@ -7,451 +7,370 @@
...
-				start = len(data) - len(remaining)
-			} else {
-				keyBlock = data[start : len(data)-len(remaining)]
-				start += len(keyBlock)
-			}
+	// Deployment
+	if co.Deployment != nil {
+		// topology options: WithSeedlist, WithURI, WithSRVServiceName, WithSRVMaxHosts, and WithServerOptions
+		// server options: WithClock and WithConnectionOptions + default maxPoolSize
+		if len(serverOpts) > 2+defaultOptions || hasOptionalData(*co) {
```
Comment (prestonvasquez) [inline x/mongo/driver/topology/topology_options.go:358]: This change is of note. The original design looked at the length of the topology options slice to see if we were over assigning a number of optional topology options. That isn't possible anymore since we have deprecated the options pattern. So, instead, we use the `hasOptionalData` function to see if the underlying client options have more than the minimum data.

File: x/mongo/driver/topology/topology_options.go [hunk 8]
```diff
@@ -7,451 +7,370 @@
...
-			c.mode = SingleMode
-		}
+type client interface {
+	GetClusterClock() *session.ClusterClock
+	SetServerAPI(*driver.ServerAPIOptions)
+}
 
-		c.seedList = cs.Hosts
+// newConfig will translate data from client options into a topology config for building non-default deployments.
+func newConfig(co *options.ClientOptions, client client) (*Config, error) {
```
Comment (matthewdale) [inline x/mongo/driver/topology/topology_options.go:67]: It may be simpler to move this `ClientOptions -> Config` logic to the `mongo` package. If you do that, you can get rid of the `client` interface because you have direct access to `GetClusterClock()` and `SetServerAPI`. Is there anything that would be made more complex or block moving this logic to the `mongo` package?

File: x/mongo/driver/topology/topology_options.go [hunk 9]
```diff
@@ -7,451 +7,370 @@
...
 
-		if cs.MaxConnIdleTime > 0 {
-			connOpts = append(connOpts, WithIdleTimeout(func(time.Duration) time.Duration { return cs.MaxConnIdleTime }))
-		}
+	cfgp := new(Config)
 
-		if cs.MaxPoolSizeSet {
-			c.serverOpts = append(c.serverOpts, WithMaxConnections(func(uint64) uint64 { return cs.MaxPoolSize }))
-		}
+	// TODO(GODRIVER-814): Add tests for topology, server, and connection related options.
```
Comment (matthewdale) [inline x/mongo/driver/topology/topology_options.go:91]: [GODRIVER-814](https://jira.mongodb.org/browse/GODRIVER-814) is closed and this comment seems outdated, so can be removed.

File: x/mongo/driver/topology/topology_options_test.go (hunk 5 above)
Comment (matthewdale) [reply x/mongo/driver/topology/topology_options_test.go:17]: If it's redundant with a `connstring` or `options` package test (or unified spec test), then it seems safe to remove.

**Review by matthewdale — COMMENTED** [review]
Great work so far!

**Review by divjotarora — DISMISSED** [review]

File: mongo/options/clientoptions.go (hunk 3 above)
Comment (prestonvasquez) [reply mongo/options/clientoptions.go:207]: @matthewdale @benjirewis What are your thoughts on fixing this bug in this ticket?

File: examples/documentation_examples/examples_test.go [hunk 10]
```diff
@@ -147,9 +146,12 @@ func TestDocumentationExamples(t *testing.T) {
 }
 
 func createTopology(mt *mtest.T) *topology.Topology {
```
[the remark this answers is not in the document]
Comment (prestonvasquez) [reply examples/documentation_examples/examples_test.go:148]: Wow, great catch!

File: x/mongo/driver/topology/topology_options.go [hunk 11]
```diff
@@ -7,451 +7,370 @@
...
-	serverOpts             []ServerOption
-	cs                     connstring.ConnString // This must not be used for any logic in topology.Topology.
-	uri                    string
-	serverSelectionTimeout time.Duration
-	serverMonitor          *event.ServerMonitor
-	srvMaxHosts            int
-	srvServiceName         string
-	loadBalanced           bool
+// Config is used to construct a topology.
+type Config struct {
```
[the remark this answers is not in the document]
Comment (prestonvasquez) [reply x/mongo/driver/topology/topology_options.go:27]: If we _only_ want users to build configs from client options, then no. I actually prefer leaving it unexported, if that's fine with the team.

File: x/mongo/driver/examples/cluster_monitoring/main.go [hunk 12]
```diff
@@ -14,7 +14,7 @@ import (
 )
 
 func main() {
```
[the remark this answers is not in the document]
Comment (prestonvasquez) [reply x/mongo/driver/examples/cluster_monitoring/main.go:16]: This is a very solid idea, this one makes sense for this PR. Should I log a ticket to do something similar to the other tests in the examples/ directory?

File: x/mongo/driver/topology/topology_options.go (hunk 8 above)
Comment (prestonvasquez) [reply x/mongo/driver/topology/topology_options.go:67]: Since the client package needs to build a deployment from the topology package and the topology tests require building configuration structs from client options, this would create a circular dependency between the `topology` package and the `mongo` package.

Divjot is suggesting we put the accessor data on the topology structures and set them on the client as needed: https://github.com/mongodb/mongo-go-driver/pull/1044#discussion_r945231313 . What are your thoughts on this approach?

File: mongo/client.go [hunk 13]
```diff
@@ -25,18 +23,15 @@ import (
...
-	"go.mongodb.org/mongo-driver/x/mongo/driver/ocsp"
 	"go.mongodb.org/mongo-driver/x/mongo/driver/operation"
 	"go.mongodb.org/mongo-driver/x/mongo/driver/session"
 	"go.mongodb.org/mongo-driver/x/mongo/driver/topology"
 )
 
 const (
-	defaultLocalThreshold        = 15 * time.Millisecond
-	defaultMaxPoolSize    uint64 = 100
+	defaultLocalThreshold = 15 * time.Millisecond
```
[the remark this answers is not in the document]
Comment (prestonvasquez) [reply mongo/client.go:34]: There are two things that make me hesitant about this:

1. `defaultLocalThreshold` is used to verify an expectation in [this](https://github.com/mongodb/mongo-go-driver/blob/master/mongo/client_test.go#L167) test.
2. The `newConfig` logic, like its predecessor, counts and [uses `defaultOptions` for error handling](https://github.com/mongodb/mongo-go-driver/blob/master/mongo/client.go#L711) .

I think this is worth filing an investigation ticket for, though.

File: x/mongo/driver/topology/topology_test.go [hunk 14]
```diff
@@ -620,16 +620,17 @@ func TestTopologyConstruction(t *testing.T) {
 			pollingRequired bool
 		}{
 			{"normal", "mongodb://localhost:27017", false},
-			{"srv", "mongodb+srv://localhost:27017", true},
+			// {"srv", "mongodb+srv://localhost:27017", true},
```
Comment (benjirewis) [inline x/mongo/driver/topology/topology_test.go:623]: Why is this commented out?

File: x/mongo/driver/topology/topology_options.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (benjirewis) [reply x/mongo/driver/topology/topology_options.go:27]: Hmm would any internal teams need to build a `Config` themselves? It's under `x`, so we can certainly modify it later if need be.

File: x/mongo/driver/topology/topology_options.go (hunk 8 above)
Comment (benjirewis) [reply x/mongo/driver/topology/topology_options.go:67]: I like the approach @divjotarora suggests if moving to the `mongo` package will cause a circular dependency.

File: x/mongo/driver/topology/topology_options.go [hunk 15]
```diff
@@ -7,451 +7,370 @@
...
-		var connOpts []ConnectionOption
-
-		if cs.AppName != "" {
-			c.serverOpts = append(c.serverOpts, WithServerAppName(func(string) string { return cs.AppName }))
-		}
-
-		if cs.Connect == connstring.SingleConnect || (cs.DirectConnectionSet && cs.DirectConnection) {
-			c.mode = SingleMode
-		}
+type client interface {
```
[the remark this answers is not in the document]
Comment (benjirewis) [reply x/mongo/driver/topology/topology_options.go:61]: I like this approach. What do you mean by "leaky" abstraction though?

File: x/mongo/driver/topology/topology_options.go [hunk 16]
```diff
@@ -7,451 +7,370 @@
...
+	GetClusterClock() *session.ClusterClock
+	SetServerAPI(*driver.ServerAPIOptions)
+}
 
-		c.seedList = cs.Hosts
+// newConfig will translate data from client options into a topology config for building non-default deployments.
+func newConfig(co *options.ClientOptions, client client) (*Config, error) {
+	const defaultMaxPoolSize = 100
+	var serverAPI *driver.ServerAPIOptions = nil
```
Comment (benjirewis) [inline x/mongo/driver/topology/topology_options.go:69]:
```suggestion
	var serverAPI *driver.ServerAPIOptions
```
Isn't specifying the zero-value here and below redundant?

File: x/mongo/driver/topology/topology_options.go (hunk 7 above)
Comment (benjirewis) [reply x/mongo/driver/topology/topology_options.go:358]: This seems like a fine alternative to me. The comment above is now (and was) a little confusing though. Can we update it to explain what exactly we're checking for?

**Review by benjirewis — CHANGES_REQUESTED** [review]
Looking great! This is a pretty nuanced code clean-up.

File: x/mongo/driver/topology/topology_options.go (hunk 8 above)
Comment (matthewdale) [reply x/mongo/driver/topology/topology_options.go:67]: Yeah, I see the source of the circular dependency now. Sounds like it's still important to be able to initialize a `Topology` from a connection string for some tests, which we'd have to stop doing if we moved this logic into the `mongo` package. Ideally any test that needs to construct a `Topology` would be able to do so without the `connstring -> Topology` logic, but that doesn't seem immediately feasible, so seems like a good idea to keep this logic in the `topology` package.

As far as the `topology.client` interface, I agree it should be removed. There are 2 functions in the `topology.client` interface which each handle a different piece of information.

## `GetClusterClock`
This can just be a parameter on `newConfig`. Callers that don't know the cluster clock can pass `nil`.

E.g.
```go
func NewConfig(co *options.ClientOptions, clock *session.ClusterClock) {
	// ...
}
```

## `SetServerAPI`
We can prevent the need to assign the `driver.ServerAPIOptions` back to the `Client` by exporting the function for converting an `options.ServerAPIOptions` -> `driver.ServerAPIOptions` (i.e. `convertToDriverAPIOptions`).

E.g.
```go
func (c *Client) configure(...) error {
	// ...
	if opts.ServerAPIOptions != nil {
		c.serverAPI = topology.ConvertToDriverAPIOptions(opts.ServerAPIOptions)
		// ...
}

 func NewConfig(co *options.ClientOptions, clock *session.ClusterClock) (*Config, error) {
	// ...
	if co.ServerAPIOptions != nil {
		serverAPI = ConvertToDriverAPIOptions(co.ServerAPIOptions)
		// ...
```

File: x/mongo/driver/topology/topology_options.go [hunk 17]
```diff
@@ -7,451 +7,370 @@
...
+// NewConfig will translate client options into a topology config for building deployments.
+func NewConfig(co *options.ClientOptions) (*Config, error) {
+	return newConfig(co, nil)
+}
 
-	return crt.Subject.String(), nil
+// NewConfigWithClient will translate client options into a topology config for building deployments. It will also use
+// a client implementation for accessing and setting unexported data as well as implementing non-configuration-specific
+// data.
+func NewConfigWithClient(co *options.ClientOptions, client client) (*Config, error) {
```
Comment (matthewdale) [inline x/mongo/driver/topology/topology_options.go:374]: This function will likely become unnecessary if we remove the `client` interface and just allow callers to pass in a nil-able `session.ClusterClock`.

File: x/mongo/driver/examples/cluster_monitoring/main.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (matthewdale) [reply x/mongo/driver/examples/cluster_monitoring/main.go]: Yeah, I think a GODRIVER ticket and maybe a TODO comment is the right way to go.

File: x/mongo/driver/topology/topology_options.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (matthewdale) [reply x/mongo/driver/topology/topology_options.go:27]: Yeah, based on better understanding of this change, it is obvious it needs to be exported. Please disregard!

File: x/mongo/driver/topology/topology_options.go (hunk 8 above)
Comment (prestonvasquez) [reply x/mongo/driver/topology/topology_options.go:67]: Setting the same data in two separate places doesn't seem quite right to me. I think Divjot's solution is the most uniform. Alternatively we could add a function to parse the data from `topology.Config` as a one off:

```go
// x/mongo/driver/topology/server_options.go
func ServerAPIFromServerOptions(opts []ServerOption) *driver.ServerAPIOptions {
  return newServerConfig(opts...).serverAPI
}
```

```go
// mongo/client.go
func NewClient(opts ...*options.ClientOptions) (*Client, error) { 
  // ...
  client.serverAPI = topology.ServerAPIFromServerOptions(cfg.ServerOpts)
  // ...
}
```

File: mongo/integration/initial_dns_seedlist_discovery_test.go [hunk 18]
```diff
@@ -175,11 +188,11 @@ func verifyConnstringOptions(mt *mtest.T, expected bson.Raw, cs connstring.ConnS
 // If the option is true, then we skip the test if the server is running without SSL enabled; if
```
[the remark this answers is not in the document]
Comment (prestonvasquez) [reply mongo/integration/initial_dns_seedlist_discovery_test.go:188]: I believe this is still relevant given L213:

```go
tlsConfig.InsecureSkipVerify = true
```

File: mongo/integration/mtest/setup.go [hunk 19]
```diff
@@ -57,69 +57,64 @@ var testContext struct {
 	serverless                  bool
 }
 
-func setupClient(cs connstring.ConnString, opts *options.ClientOptions) (*mongo.Client, error) {
+func setupClientAlt(opts *options.ClientOptions) (*mongo.Client, error) {
```
[the remark this answers is not in the document]
Comment (prestonvasquez) [reply mongo/integration/mtest/setup.go:60]: Ah it doesn't mean anything, I was coding with both functions for testing purposes and forgot to rename the new one!

File: x/mongo/driver/topology/polling_srv_records_test.go [hunk 20]
```diff
@@ -257,21 +253,17 @@ func TestPollingSRVRecordsLoadBalanced(t *testing.T) {
...
-		topo, err := New(
-			WithConnString(func(connstring.ConnString) connstring.ConnString { return cs }),
-			WithURI(func(string) string { return cs.Original }),
-		)
-		assert.Nil(t, err, "topology.New error: %v", err)
 		return topo
 	}
 
 	t.Run("pollingRequired is set to false", func(t *testing.T) {
-		topo := createLBTopology(t, "mongodb+srv://test1.test.build.10gen.cc/?heartbeatFrequencyMS=100")
```
[the remark this answers is not in the document]
Comment (prestonvasquez) [reply x/mongo/driver/topology/polling_srv_records_test.go:274]: `test1` is for multiple hosts which results in a load balancing error.

File: mongo/integration/mtest/setup.go [hunk 21]
```diff
@@ -57,69 +57,64 @@ var testContext struct {
...
+
 	testContext.dataLake = os.Getenv("ATLAS_DATA_LAKE_INTEGRATION_TEST") == "true"
 	testContext.requireAPIVersion = os.Getenv("REQUIRE_API_VERSION") == "true"
 
-	connectionOpts := []topology.ConnectionOption{
-		topology.WithOCSPCache(func(ocsp.Cache) ocsp.Cache {
-			return ocsp.NewCache()
-		}),
+	clientOpts := options.Client().ApplyURI(uri)
+	if testContext.requireAPIVersion {
```
[the remark this answers is not in the document]
Comment (prestonvasquez) [reply mongo/integration/mtest/setup.go:99]: Nice!

File: x/mongo/driver/topology/topology_test.go (hunk 14 above)
Comment (prestonvasquez) [reply x/mongo/driver/topology/topology_test.go:623]: This is an [invalid test](https://github.com/mongodb/specifications/blob/master/source/initial-dns-seedlist-discovery/initial-dns-seedlist-discovery.rst#:~:text=It%20is%20an%20error%20to%20specify%20a%20port%20in%20a%20connection%20string%20with%20the%20mongodb%2Bsrv%20protocol%2C%20and%20the%20driver%20MUST%20raise%20a%20parse%20error%20and%20MUST%20NOT%20do%20DNS%20resolution%20or%20contact%20hosts.) that will result in this error:

```
topology_test.go:628: error constructing topology config: error parsing uri: URI with srv must not include a port number
```

File: mongo/options/clientoptions.go (hunk 3 above)
Comment (benjirewis) [reply mongo/options/clientoptions.go:207]: On second thought, @matthewdale is probably right; I don't see where else we would merge validation logic for `ApplyURI` and `WithConnString` (or did you actually mean `WithURI`, @matthewdale). GODRIVER-1714 should probably be done here.

File: mongo/integration/initial_dns_seedlist_discovery_test.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (benjirewis) [reply mongo/integration/initial_dns_seedlist_discovery_test.go:188]: Ah I meant the next sentence, sorry.

> If the server is running with SSL enabled, then we manually set the necessary SSL options in the connection string.

Should maybe be something like

> If the server is running with SSL enabled, then we return a tls.Config with InsecureSkipVerify set to true.

File: mongo/integration/mtest/setup.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (benjirewis) [reply mongo/integration/mtest/setup.go:60]: Gotcha, looks like it's still `Alt` though.

File: x/mongo/driver/examples/server_monitoring/main.go [hunk 22]
```diff
@@ -16,6 +16,8 @@ import (
 	"go.mongodb.org/mongo-driver/x/mongo/driver/topology"
 )
 
+// TODO(GODRIVER-2527): convert to a Go doc example.
+
```
Comment (benjirewis) [inline x/mongo/driver/examples/server_monitoring/main.go:20]:
```suggestion
```

File: x/mongo/driver/examples/workload/main.go [hunk 23]
```diff
@@ -20,17 +20,19 @@ import (
...
-	"go.mongodb.org/mongo-driver/x/mongo/driver/connstring"
 	"go.mongodb.org/mongo-driver/x/mongo/driver/operation"
 	"go.mongodb.org/mongo-driver/x/mongo/driver/topology"
 )
 
 var concurrency = flag.Int("concurrency", 24, "how much concurrency should be used")
 var ns = flag.String("namespace", "test.foo", "the namespace to use for test data")
 
+// TODO(GODRIVER-2527): convert to a Go doc example.
+
```
Comment (benjirewis) [inline x/mongo/driver/examples/workload/main.go:35]:
```suggestion
```

File: x/mongo/driver/topology/topology_options.go [hunk 24]
```diff
@@ -7,451 +7,345 @@
...
-	for _, opt := range opts {
-		if opt == nil {
-			continue
-		}
-		err := opt(cfg)
-		if err != nil {
-			return nil, err
-		}
+// ConvertToDriverAPIOptions converts a options.ServerAPIOptions instance to a driver.ServerAPIOptions.
+func ConvertToDriverAPIOptions(s *options.ServerAPIOptions) *driver.ServerAPIOptions {
```
Comment (matthewdale) [inline x/mongo/driver/topology/topology_options.go:50]: Does function this need to be exported? If not, change it to be un-exported.

File: x/mongo/driver/topology/topology_options.go [hunk 25]
```diff
@@ -7,451 +7,345 @@
...
-	crt, err := x509.ParseCertificate(certDecodedBlock)
-	if err != nil {
-		return "", err
+	cfgp.ServerOpts = serverOpts
+
+	// Deployment
+	if co.Deployment != nil {
+		// A deployment cannot be created with server or topology options. To extend the optional requirements
+		// add topology options to the "hasOptionalData" function.
+		if len(serverOpts) > 2+defaultOptions || hasOptionalData(*co) {
```
Comment (matthewdale) [inline x/mongo/driver/topology/topology_options.go:345]: We're not actually doing anything with the custom deployment here, so the custom deployment check seems out-of-place. We should keep the custom deployment check logic in `mongo.NewClient`.

Another option is to remove the "custom deployment with server/topology options" check. This is intended as a "don't shoot yourself in the foot" check, but has historically been buggy and using a custom deployment is already explicitly recommended against. Should we replace this check with a note that says "Server and topology options are not honored if a custom deployment is used"?

File: mongo/client.go [hunk 26]
```diff
@@ -358,243 +358,21 @@ func (c *Client) endSessions(ctx context.Context) {
 }
 
 func (c *Client) configure(opts *options.ClientOptions) error {
```
Comment (matthewdale) [inline mongo/client.go:360]: Optional: Now that the amount of code in `configure` has been significantly reduced, consider merging the remaining code into `NewClient`.

File: x/mongo/driver/topology/topology_options.go (hunk 25 above)
Comment (prestonvasquez) [reply x/mongo/driver/topology/topology_options.go:345]: I think if we have concerns about keeping `hasOptionalData` in sync with the logic, then it's probably best to just remove this branch.

File: mongo/options/clientoptions.go (hunk 3 above)
Comment (matthewdale) [reply mongo/options/clientoptions.go:207]: Yeah, I meant `WithConnString`. Updated my comment 👍

File: x/mongo/driver/topology/topology_options.go (hunk 25 above)
Comment (matthewdale) [reply x/mongo/driver/topology/topology_options.go:345]: That sounds good to me. @benjirewis do you have any reservations about removing the "custom deployment with server/topology options" check?

The change that originally introduced that logic is https://github.com/mongodb/mongo-go-driver/commit/bb9530df7191054d68c7879dc4a2045eb71ad856. I don't see anything in the associated Jira ticket that specifically requires that check, so it seems to be a "nice to have".

**Review by matthewdale — APPROVED** [review]
Looks good 👍

File: x/mongo/driver/topology/topology_options.go (hunk 24 above)
Comment (benjirewis) [reply x/mongo/driver/topology/topology_options.go:39]: Looks like it's called from other packages, so I think it probably needs to be exported.

File: mongo/client.go [hunk 27]
```diff
@@ -136,13 +131,83 @@ func NewClient(opts ...*options.ClientOptions) (*Client, error) {
...
+	// Set default options
+	if clientOpt.MaxPoolSize == nil {
+		clientOpt.SetMaxPoolSize(defaultMaxPoolSize)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	cfg, err := topology.NewConfig(clientOpt, client.clock)
```
Comment (benjirewis) [inline mongo/client.go:203]: Does calling `NewConfig` here instead of just `configure` satisfy GODRIVER-1714 (and, if so can we make sure the final commit message contains GODRIVER-1714)? Seems like we're now doing the same validation for creating a client from a URI as we do for creating a client from `ClientOption`s.

File: x/mongo/driver/topology/topology_options.go (hunk 25 above)
Comment (benjirewis) [reply x/mongo/driver/topology/topology_options.go]: Great point; I think it's fine to remove as long as we add in a comment like what you suggested @matthewdale ("Server and topology options are not honored if a custom deployment is used").

File: x/mongo/driver/topology/topology_options.go [hunk 28]
```diff
@@ -7,451 +7,345 @@
...
+	LoadBalanced           bool
 }
 
-func newConfig(opts ...Option) (*config, error) {
-	cfg := &config{
-		seedList:               []string{"localhost:27017"},
-		serverSelectionTimeout: 30 * time.Second,
-	}
+func hasOptionalData(co options.ClientOptions) bool {
+	return co.Direct != nil && *co.Direct ||
```
[the remark this answers is not in the document]
Comment (benjirewis) [reply x/mongo/driver/topology/topology_options.go]: I think _removing_ the custom deployment check for extra server and topology options is fine as long as we document. (No need to move any logic to `mongo` in that case, I guess).

**Review by benjirewis — CHANGES_REQUESTED** [review]
Almost done I think! Let's make sure we document that server + topology options may not be honored when a custom deployment is used if we agree to remove that check. ~Looks like static analysis is still failing too, so let's investigate that before merging.~ Seems like static analysis failures will be fixed in #1055.

**Comment by prestonvasquez** [comment]
> Almost done I think! Let's make sure we document that server + topology options may not be honored when a custom deployment is used if we agree to remove that check. ~Looks like static analysis is still failing too, so let's investigate that before merging.~ Seems like static analysis failures will be fixed in #1055.

@benjirewis This should already be documented [here](https://github.com/mongodb/mongo-go-driver/pull/1044/files#diff-e3ed760c474d9d11f91ebd6bfd887479ab4921538331a6877792365274c8b4d7R51)

File: mongo/client.go (hunk 27 above)
Comment (prestonvasquez) [reply mongo/client.go:203]: GODRIVER-1714 is not resolved by this. `topology.NewConfig` deprecates the `tolopology.WithConnString` while merging most of the logic of configure into the topology package. However, a user could build a URI without using the connstring package and then pass it to the client options with `ClientOptions.ApplyURI`, this would bypass these validations. GODRIVER-1714 would require merging the validations from `connstring.ParseAndValidate` into `ClientOptions.Validate`.

**Review by benjirewis — APPROVED** [review]
Fantastic work!
